### PR TITLE
Fix `cache_position` initialisation for generation with `use_cache=False`

### LINKED ISSUE #30482: `generate` method does not work with `use_cache=False` in `Llama-2` model with `model.config._attn_implementation = "eager"`
@ArthurZucker , @gante 

The issue still persists when `model.config._attn_implementation = "eager"` is used. Currently it works because the implementation relies on `sdpa` attention implementation, which does not use `causal_mask`. 
```py
class LlamaModel(LlamaPreTrainedModel):
    ...

    def _update_causal_mask(
        self,
        attention_mask: torch.Tensor,
        input_tensor: torch.Tensor,
        cache_position: torch.Tensor,
        past_seen_tokens: int,
    ):
        ...

        if self.config._attn_implementation == "sdpa":
            # For SDPA, when possible, we will rely on its `is_causal` argument instead of its `attn_mask` argument,
            # in order to dispatch on Flash Attention 2.
            if AttentionMaskConverter._ignore_causal_mask_sdpa(
                attention_mask, inputs_embeds=input_tensor, past_key_values_length=past_seen_tokens
            ):
                return None
```

Ideally, the code should work with any attention implementation. I believe the problem is that the `cache_position` is wrong when we use `model.generate` with `use_cache=False`.

_Originally posted by @nurlanov-zh in https://github.com/huggingface/transformers/issues/30347#issuecomment-2076917488_

### PR DESCRIPTION
**Pull request by nurlanov-zh**
# What does this PR do?
Fixes #30482

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@ArthurZucker @younesbelkada @gante

### CODE REVIEW

File: src/transformers/generation/utils.py [hunk 1]
```diff
@@ -1297,6 +1297,10 @@ def _prepare_generation_config(
 
     def _get_initial_cache_position(self, input_ids, model_kwargs):
         """Calculates `cache_position` for the pre-fill stage based on `input_ids` and optionally past length"""
+        if "use_cache" in model_kwargs and not model_kwargs["use_cache"]:
```
Comment (ArthurZucker) [inline src/transformers/generation/utils.py:1300]:
```suggestion
        if not model_kwargs.get("use_cache", True):
```

Comment (ArthurZucker) [reply src/transformers/generation/utils.py:1300]: If you are not using the cache, you should still pass the cache positions, but they should be automatically created using the shape of the input ids. cc @gante as well!

**Review by ArthurZucker — COMMENTED** [review]
Thanks! Let's maybe avoid relying on the model's cache position initialization !

File: src/transformers/generation/utils.py (hunk 1 above)
Comment (nurlanov-zh) [reply src/transformers/generation/utils.py]: Included the `use_cache` check into `_update_model_kwargs_for_generation` function.

Comment (zucchini-nlp) [reply src/transformers/generation/utils.py]: It seems to me that [these](https://github.com/huggingface/transformers/blob/1e05671d213642a67dd418c13a18a17a18c25117/src/transformers/models/llama/modeling_llama.py#L1084-L1086) lines should be an "if...else", which will solve the problem with caching. 

Because if sequence length is not 1, making a diagonal mask will be enough, and the `causal_mask *= torch.arange(target_length, device=device) > cache_position.reshape(-1, 1)` is unnecessarily unmasking all the positions now.

Comment (nurlanov-zh) [reply src/transformers/generation/utils.py]: @zucchini-nlp , yes, using `cache_position` when `use_cache=False` does not make much sense. But with the proposed fixes, `cache_position` is None, and it is initialized in the model's forward pass as `cache_position = torch.arange(past_seen_tokens, past_seen_tokens + inputs_embeds.shape[1], device=inputs_embeds.device)`, therefore it does not influence the update of `causal_mask`.

**Review by gante — APPROVED** [review]
LGTM, thank you for the fix 👍 

@ArthurZucker tbh, when we have `use_cache=False`, we shouldn't even be creating or using `cache_positions` at all (as the name of the variable indicates, its purpose is cache-related). ATM, the only dependency is the causal mask creation function, which uses `cache_positions` as a proxy for the sequence length (i.e. we can easily rearrange the code to avoid using it when there is no cache :) )

**Review by ArthurZucker — APPROVED** [review]
Yes @gante, agreed would split both paths a bit better

**Comment by ArthurZucker** [comment]
Thanks 🤗
